### ISSUE TITLE
[API Proposal]: Add sliding expiration to `HybridCacheEntryOptions`

### ISSUE THREAD

**Issue by CoryCharlton** [dotnet/extensions#5649]
### Background and motivation

This is a duplicate of https://github.com/dotnet/aspnetcore/issues/56754 which I believe is in the wrong repo.

---
Are there any plans to implement a sliding expiration for `HybridCacheEntryOptions`?

The existing `DistributedCacheEntryOptions` has a `SlidingExpiration` property which has been working well for my team and I. We have another mechanism to invalidate cache keys when updates to the underlying data occurs so we like to keep the data in Redis for as long as it is being used and eventually expire when that particular data is not regularly accessed.

### Desired Solution

Ideally, the same `SlidingExpiration` property would be added to `HybridCacheEntryOptions` and the value would be passed along in `ToDistributedCacheEntryOptions` and handled in the downstream `IDistributedCache` calls.
---

### API Proposal

```csharp
namespace Microsoft.Extensions.Caching.Hybrid;

public sealed class HybridCacheEntryOptions
{
    public TimeSpan? SlidingExpiration { get; set; }
}
```

### API Usage

```csharp
// Fancy the value
var c = new MyFancyCollection<int>();
c.Fancy(42);

// Getting the values out
foreach (var v in c)
    Console.WriteLine(v);
```

### Risks

At first glance the risk involved here is implementing this in a manner that does not negate the performance benefits provided by using `HybridCache`.

For instance if a cache request hits the local `MemoryCache` it's expiration would get updated but the entry in the `IDistributedCache` would not.

**Comment by mvxproject** [comment]
In the case when we get the cache from `MemoryCache`  we can refresh the distributed cache before returning the value.

`_backendCache.RefreshAsync(key, token);`

**Comment by bill-poole** [comment]
There would be value in using sliding expiration for the distributed cache and absolute expiration for the local cache, setting the expiry for entries in the distributed cache slightly longer than those in the local cache. This would keep the regularly accessed entries in the distributed cache live because the entries in the local cache would expire before the corresponding entries in the distributed cache, causing those entries to be accessed in the distributed cache, pushing their expiration back.

This would reduce load on the underlying database by significantly reducing requests for regularly accessed items.

**Comment by bill-poole** [comment]
> how often do we trigger RefreshAsync when fetching something from L1 - is it every time? every N times? every N seconds?

If using Redis pipelining, then it is possible to send an EXPIRE and GET in one roundtrip, thus pushing back the expiration on each access. Not sure about other cache implementations though. But if an implementation supports this, then it could possibly be done by way of a configuration setting on the `IDistributedCache` implementation that specifies whether to reset the expiration on each access. That way there would be no need to explicitly invoke `RefreshAsync`.

**Comment by mgravell** [comment]
> If using Redis pipelining, then it is possible to send an EXPIRE and GET in one roundtrip,

That misses the point; the backend fetch get should *already* trigger sliding expiration (although it is a little more nuanced than blind pipelining in the case of Redis) - however, the entire point of HybridCache is that we *don't* use the distributed-cache every call, so we need some additional for L1 hits to cause L2 refreshes *at some rate* than isn't necessarily "every L1 hit causes an L2 refresh".

**Comment by bill-poole** [comment]
If local (L1) has absolute expiration, say one minute, and the distributed cache (L2) has a sliding window of 2 minutes that is refreshed on each access, then the distributed cache will only be accessed once per minute (assuming a very high rate of accesses), sliding the accessed entry's expiration back one minute at that time. When the L1 cache expires (after one minute), the next access will go to the L2 cache, sliding the L2 expiration back another minute, and the retrieved entry will then be added to the L1 cache, and all subsequent accesses will go to the L1 cache until that entry expires one minute later.

If each L2 access slides the expiration window back, then there is no need to call `RefreshAsync` on each access, every N times, nor every N seconds.

Apologies if I've misunderstood something here.

**Comment by mgravell** [comment]
If I may rephrase that, @bill-poole , to check my understanding:

- add sliding expiration onto HC options
- assume/document that L1 (MC) duration should be less than the sliding expiration
- have sliding expiration *only* impact L2 (DC)
- have the backend sliding expiration be renewed simply via L1 expiration and re-fetch

Is that right? I admit, it has the advantage of being very simple to understand and implement - the whole thing would be a 1 line change to the code that creates DC options from HC options.

**Comment by bill-poole** [comment]
Yes, that's right. Sliding expiration would apply only to L2 (DC), while L1 (HC) would use absolute expiration with a duration less than the sliding expiration for L2 (DC). Sliding expiration for L2 (DC) only makes sense if accesses are propagated through L1 (MC) often enough to keep the cache entries live, which doesn't happen if L1 (MC) uses sliding expiration.

L1 (MC) then serves to significantly reduce load on L2 (DC), while drip feeding requests to L2 (DC) for those entries being regularly accessed to keep those entries live in L2 (DC) for as long as those entries are regularly accessed.

**Comment by jodydonetti** [comment]
Honest question, since I've never used sliding expiration that much because I typically use a different approach: does the RedisCache impl support sliding expiration? I don't mean the RefreshAsync method, but the use of `DistributedCacheEntryOptions.SlidingExpiration`.
I never looked into it further how the use of that entry option translates to a Redis command.

**Comment by jodydonetti** [comment]
Hi @mgravell 

> Is that right? I admit, it has the advantage of being very simple to understand and implement - the whole thing would be a 1 line change to the code that creates DC options from HC options.

Watch out for users not using L2, either because not setup at all or because of some entry options flags per-call: what would happen then? No L1 sliding? Automatic L1 sliding when no L2?

**Comment by mgravell** [comment]
> does the RedisCache impl support sliding expiration

Yes. All handled automatically inside the "get" APIs.

> No L1 sliding? Automatic L1 sliding when no L2?

Great question. I'd assume the latter, i.e. L1 sliding if no L2.

**Comment by mvxproject** [comment]
`when fetching something from L1 - is it every time? every N times? every N seconds? `

I think, we must call **RefreshAsync** every time we  fetch something from L1,

Certainly we lose the performance of the **HybridCache**, but we gain more compatibility,

if we need the power of **hybridcahe**, we don't specify L2 sliding.

**Comment by mgravell** [comment]
(deleted by poster)

> I think, we must call RefreshAsync every time we fetch something from L1,
>
> Certainly we lose the performance of the HybridCache, but we gain more compatibility,
>
> if we need the power of hybridcahe, we don't specify L2 sliding.

Responding for completeness. I'm kinda reluctant to call down *every* hit; IMO if we did that:

- it should be opt in
- we should defer and buffer for *some delay* - which could be short

For example, a `TimeSpan?` property that, if specified triggers something (during L1 hit) on `CacheItem`:

- if not yet active, a new callback with that timespan is registered; on callback: clear the flag and call `RefreshAsync`
- if already active, do nothing

Meaning: you set the delay for 20 seconds; the first time it is accessed, the 20 second delay is triggered; any other hits in those 20 seconds do nothing; after 20 seconds we hit the refresh path and start afresh for any new hits.

**Comment by mvxproject** [comment]
> Meaning: you set the delay for 20 seconds; the first time it is accessed, the 20 second delay is triggered;

@mgravell , What if the L2 sliding expires before the 20s TimeSpan ?

**Comment by mgravell** [comment]
If your cache is so short-lived that 20 seconds is likely to be a problems, then choosing 20 seconds was unwise. However, generalising: we're trading the very small chance of false expirations for a significant reduction in the backend overhead. If you are happy to pay a larger overhead for smaller (or even zero - I'm guessing we'd make TimeSpan.Zero mean "call every time") chance of false expirations: great, do that!

**Comment by mvxproject** [comment]
This is an example of the case I was mentioning:

- first hit: we create L1 and L2 entries (L2 Sliding 5min)
- second hit after 4min 50: we get the value from L1 and Set TimeSpan 20s but the L2 Sliding has 10s left.

**Comment by mgravell** [comment]
Well, the other way of spinning the delay is to keep a "last refreshed" (equally: "don't refresh until" - same logic) token; so another entirely valid sequence is:

- query comes in, miss in L1, hit from L2, stored in L1 (implicit refresh, set "last refreshed" to now)
- (time passes)
- query or queries comes in, hit from L1, but inside "last refreshed + delta" window; do nothing
- query comes in, hit from L1, outside "last refreshed + delta" window, so: update "last refreshed" and issue async refresh
- query comes in, hit from L1, now inside blah, do nothing

In your example above, this would mean we *do* perform a refresh at 0:04:50 (or any time after 0:00:20) - that work?
